// code-narrator/dist/src/documentation/plugins/builders/AnalyzeProject/AnalyzeProjectBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseBuilder_1 = __importDefault(require("../BaseBuilder"));
class AnalyzeProjectBuilder extends BaseBuilder_1.default {
    constructor() {
        super('AnalyzeProject');
    }
    async generate() {
        /*
        let document = DocumentationCache.getProjectInfo()
        if (document) return;

        let project_path = App.Project.project_path
        let entry_file = ConfigHelper.get('entry_file');
        const entryFile = FileStructure.getContent(path.join(project_path, entry_file));
        if (entryFile == '') return;

        let args = {entryFile};
        await super.generateDocumentationAndCache({
            args,
            name: "Behaviour",
            pathToFile: './',
            folderPath: './',
            sidebarPosition: 2,
            sidebarLabel: 'Behaviour'
        })*/
    }
    async render(document) {
        return document.documentation;
    }
}
exports.default = AnalyzeProjectBuilder;
